### PR TITLE
🐛 Set imported binding as an alias of target symbol

### LINKED ISSUE #1154: References to use statements aren't being resolved properly by mcdoc yet
image
image
image

The use statement itself is properly resolved to the definition, but when referencing it somewhere, you get a reference with the same name but the path of the current module.

This eventually leads to `options.context.symbols.query(ctx.doc, 'mcdoc', '::java::data::predicate::Predicate')` being resolved as a union of `(::java::data::predicate::LootCondition | [::java::data::predicate::LootCondition])` which itself is not a valid mcdoc reference.

```
use super::loot::LootCondition

type Predicate = (LootCondition | [LootCondition])
dispatch minecraft:resource[predicate] to Predicate
```

### PR DESCRIPTION
**Pull request by SPGoding**
Fixes #1154.

https://github.com/SpyglassMC/Spyglass/assets/15277496/221a79c2-a43b-4c87-8e4e-bed2a97012af

The file-level binding introduced by `use` is now treated as an alias of the global symbol.

### CODE REVIEW

File: packages/core/src/symbol/SymbolUtil.ts [hunk 1]
```diff
@@ -972,7 +972,7 @@ type QueryMemberCallback = (this: void, query: SymbolQuery) => unknown
 /* istanbul ignore next */
 export class SymbolQuery {
 	readonly category: string
-	readonly path: readonly string[]
+	path: readonly string[]
 	readonly #doc: TextDocument
```
Comment (TheAfroOfDoom) [inline packages/core/src/symbol/SymbolUtil.ts:976]: what does a hash `#` prefix indicate for class properties?

File: packages/mcdoc/src/binder/index.ts [hunk 2]
```diff
@@ -526,16 +539,22 @@ async function bindPath(
...
+				try {
+					query.resolveAlias().enter({
+						usage: {
+							type: 'reference',
+							node: identNode,
+							fullRange: node,
+							skipRenaming: LiteralNode.is(identNode),
+						},
+					})
+				} catch (_ignored) {
```
Comment (TheAfroOfDoom) [inline packages/mcdoc/src/binder/index.ts:552]: (nitpick)

```suggestion
				} catch {
```

if we're already explaining why we do nothing with the error i don't see the need to grab the parameter

File: packages/mcdoc/src/binder/index.ts [hunk 3]
```diff
@@ -247,10 +247,15 @@ function hoist(node: ModuleNode, ctx: McdocBinderContext): void {
...
 		// hoistUseStatement associates the AST node with the binding definition in the file symbol table,
-		// which may get overridden by bindUseStatement in the later stage as an reference to the imported symbol in the global symbol table.
+		// which will get overridden by bindUseStatement in the later stage as an reference to the imported symbol in the global symbol table.
 		// This way when the user tries to go to definition on the path in the use statement,
 		// they will go to the definition in the imported file.
 
+		/**
+		 * String representation of the absolute path of the imported symbol.
+		 */
+		const importedPathString = pathArrayToString(resolvePath(path, ctx))
```
Comment (TheAfroOfDoom) [inline packages/mcdoc/src/binder/index.ts:257]: (nitpick)

this fits comfortably on one line if you're open to that stylistically, and still renders as JSDoc

```suggestion
		/** String representation of the absolute path of the imported symbol. */
		const importedPathString = pathArrayToString(resolvePath(path, ctx))
```

**Comment by NeunEinser** [comment]
This does not fully fix #1154.

The symbol deployed to mcdoc still contains the non-aliased reference. If I delete the `visibility: SymbolVisibility.File,` in line 271, and edit my code to also do `symbol.resolveAlias()` when performing the lookup, I _do_ get the correct node with the correct typedef.

I am not sure it is desirable to declare a public alias here though, I think I would prefer it if the reference type used the resolved path, and the alias wasn't public.

**Review by NeunEinser — CHANGES_REQUESTED** [review]
This does not fully fix #1154.

The symbol deployed to mcdoc still contains the non-aliased reference. If I delete the `visibility: SymbolVisibility.File,` in line 271, and edit my code to also do `symbol.resolveAlias()` when performing the lookup, I _do_ get the correct node with the correct typedef.

I am not sure it is desirable to declare a public alias here though, I think I would prefer it if the reference type used the resolved path, and the alias wasn't public.

File: packages/core/src/symbol/SymbolUtil.ts (hunk 1 above)
Comment (SPGoding) [reply packages/core/src/symbol/SymbolUtil.ts:976]: [private property](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties)

File: packages/mcdoc/src/binder/index.ts (hunk 2 above)
Comment (SPGoding) [reply packages/mcdoc/src/binder/index.ts:552]: didn't know you could do that

File: packages/mcdoc/src/binder/index.ts (hunk 3 above)
Comment (MulverineX) [reply packages/mcdoc/src/binder/index.ts:257]: it makes more sense to consistently have JSDoc formatted this way though, and having a line break in-between is nice for readability, since the comment color is so dark

Comment (NeunEinser) [reply packages/mcdoc/src/binder/index.ts:257]: Yeah, I think I also prefer to have it always multiline.

Comment (TheAfroOfDoom) [reply packages/mcdoc/src/binder/index.ts:257]: works for me!

**Comment by SPGoding** [comment]
> The symbol deployed to mcdoc still contains the non-aliased reference.

I think that's solved now. I also added [tests](https://github.com/SpyglassMC/Spyglass/pull/1193/files#diff-74b13033e73582aa8ac1f6d0c1c2eeb89728f4a8cd2925e8eb9b20fbcee9938eR95) that shows the global target path is stored in the symbol table instead of the file-level alias path.

**Review by NeunEinser — APPROVED** [review]
Nice, this works like a charm and no cluttering the global symbol table with import aliases =)

**Comment by NeunEinser** [comment]
Btw, you can use my [NeunEinser:neun](https://github.com/NeunEinser/Spyglass/tree/neun) branch which is a combination of SPG's branch and my mcdoc branch for testing and compare errors with just the [NeunEinser:mcdoc-validator](https://github.com/NeunEinser/Spyglass/tree/mcdoc-validator) branch on its own to see it working.

**Review by NeunEinser — APPROVED** [review]
I don't think I quite understand why alias didn't work. Is this a bug in how aliases are implemented or expected behavior when trying to declare an alias of sth not loaded yet?


If it is expected behavior, how would you use aliases in a way that avoids this problem?

All known edge cases are solved, [NeunEinser:neun](https://github.com/NeunEinser/Spyglass/tree/neun) is updated.

**Comment by SPGoding** [comment]
> expected behavior when trying to declare an alias of sth not loaded yet

it's mostly this. i could still use an alias but don't actually check the existence of the target when resolving it, in which case using  a custom data is just simpler than using an alias.

File: packages/core/src/symbol/SymbolUtil.ts [hunk 4]
```diff
@@ -1337,6 +1337,8 @@ export class SymbolQuery {
 			}
 			this.#symbol = result
 			this.#map = result.parentMap
+			this.#parentSymbol = result.parentSymbol
+			this.path = result.path
```
Comment (misode) [inline packages/core/src/symbol/SymbolUtil.ts:1341]: Since aliases are no longer used in this specific case, are these changes still needed? Or was this an issue that always existed before that you discovered while trying aliases?

**Review by misode — APPROVED** [review]
The issues on Neun's branch seem to be resolved!

File: packages/core/src/symbol/SymbolUtil.ts (hunk 4 above)
Comment (SPGoding) [reply packages/core/src/symbol/SymbolUtil.ts:1341]: It's an issue that always existed
